torch/_inductor/select_algorithm.py

```diff
@@ -1066,63 +1113,177 @@ def _jinja2_env():
+        if cache_key is None:
+            return None
+
+        entry = self._cache.get(cache_key, None)
+        return entry
+
+    def put_entry(
+        self,
+        cache_key: Optional[str],
+        code: str,
+        extra: str,
+        events: list[Any],
+    ) -> None:
+        if cache_key is None:
+            return
+        entry = GeneratedCodeCacheEntry(code, extra, events)
+        self._cache.update({cache_key: entry})
+
+
 class TritonTemplate(KernelTemplate):
     # Allow subclasses to override the kernel type
     kernel_type: type[Any] = TritonTemplateKernel
     index_counter = itertools.count()
     all_templates: dict[str, "TritonTemplate"] = {}
 
-    def __init__(self, name: str, grid: Any, source: str, debug=False) -> None:
+    def __init__(
+        self,
+        name: str,
+        grid: Any,
+        source: str,
+        debug=False,
+    ) -> None:
         super().__init__(name)
         self.grid = grid
         self.template = self._template_from_string(source)
         assert name not in self.all_templates, "duplicate template name"
-        self.all_templates[name] = self
+        TritonTemplate.all_templates[name] = self
         self.debug = debug
+        self._generated_code_cache: GeneratedCodeCache = GeneratedCodeCache()
+        clear_on_fresh_inductor_cache(self._generated_code_cache)
 
-    def generate(  # type: ignore[override]
+    # These class fields are used for testing _generated_code_cache.
+    # When this flag is on, we ensure that the cached results and the generated result if cache
+    # was not used are the same.
+    test_cache = False
+    generated_module_cache_hit = 0
+    generated_module_cache_miss = 0
+
+    def generate_and_load(
         self,
-        input_nodes,
-        layout,
-        num_stages,
-        num_warps,
-        num_consumer_groups=0,
-        num_buffers_warp_spec=0,
-        prefix_args=0,
-        suffix_args=0,
-        epilogue_fn=identity,
-        subgraphs=None,
-        mutated_inputs=None,
-        call_sizes=None,
-        workspace_arg: Optional[WorkspaceArg] = None,
-        **kwargs,
-    ):
-        """This function generates a TritonTemplateCaller
+        input_nodes: tuple[ir.StorageBox],
+        num_stages: int,
+        num_warps: int,
+        call_sizes: list[sympy.core.symbol.Symbol],
+        prefix_args: int,
+        suffix_args: int,
+        epilogue_fn: Optional[Callable[..., Any]],
+        subgraphs: Optional[list[ir.Buffer]],
+        workspace_arg: Optional[WorkspaceArg],
+        num_consumer_groups: int,
+        num_buffers_warp_spec: int,
+        layout: ir.Layout,
+        kwargs: dict[str, Any],
+    ) -> Optional[
+        tuple[ModuleType, str, tuple[str, ...], OrderedSet[str], Any, dict[str, Any]]
+    ]:
+        """Generate the python code and load it into the current process"""
+        cache_key = self._generated_code_cache.make_key(
```

Comment:
This part is rather hard to review.. would it be possible to stack it, and in the bottom pr always cache miss ?

Reply:
I am not sure how easy that would be at this moment i will loop at splitting at the end after i address all comments